File: TsinghuaNetHelper/BackgroundHelper.cpp
#include "pch.h"

#include "BackgroundHelper.h"
#include "winrt/Windows.ApplicationModel.Background.h"
#include "winrt/Windows.UI.Xaml.Interop.h"

using namespace std;
using namespace linq;
using namespace winrt;
using namespace Windows::ApplicationModel::Background;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml::Interop;

namespace winrt::TsinghuaNetHelper::implementation
{
    IAsyncOperation<bool> BackgroundHelper::RequestAccessAsync()
    {
        auto status{ co_await BackgroundExecutionManager::RequestAccessAsync() };
        if (status == BackgroundAccessStatus::Unspecified ||
            status == BackgroundAccessStatus::DeniedByUser ||
            status == BackgroundAccessStatus::DeniedBySystemPolicy)
        {
            co_return false;
        }
        co_return true;
    }

    void UnregisterTask(wstring_view const& name)
    {
        auto ts{ BackgroundTaskRegistration::AllTasks() };
        auto f{ ts.First() };

        for (auto r : BackgroundTaskRegistration::AllTasks() >>
                          select([](auto t) { return t.Value(); }) >>
                          where([&](auto r) { return r.Name() == name; }))
        {
            r.Unregister(true);
        }
    }

    BackgroundTaskBuilder NewTask(wstring_view const& name, TypeName const& type)
    {
        BackgroundTaskBuilder task;
        task.Name(name);
        task.TaskEntryPoint(type.Name);
        return task;
    }

    constexpr wstring_view LIVETILETASK{ L"LIVETILETASK" };
    constexpr wstring_view LOGINTASK{ L"LOGINTASK" };

    void BackgroundHelper::RegisterLiveTile(bool reg)
    {
        UnregisterTask(LIVETILETASK);
        if (reg)
        {
            auto task{ NewTask(LIVETILETASK, xaml_typename<LiveTileTask>()) };
            task.SetTrigger(TimeTrigger(15, false));
            task.AddCondition(SystemCondition(SystemConditionType::InternetAvailable));
            task.Register();
        }
    }

    void BackgroundHelper::RegisterLogin(bool reg)
    {
        UnregisterTask(LOGINTASK);
        if (reg)
        {
            auto task{ NewTask(LOGINTASK, xaml_typename<LoginTask>()) };
            task.SetTrigger(SystemTrigger(SystemTriggerType::NetworkStateChange, false));
            task.Register();
        }
    }
} // namespace winrt::TsinghuaNetHelper::implementation

File: TsinghuaNetUWP.Background/BackgroundHelper.vb
Imports Windows.ApplicationModel.Background

Public Module BackgroundHelper
    Private Async Function RequestAccessImplAsync() As Task(Of Boolean)
        Dim status = Await BackgroundExecutionManager.RequestAccessAsync()
        Return Not (
            status = BackgroundAccessStatus.Unspecified OrElse
            status = BackgroundAccessStatus.DeniedByUser OrElse
            status = BackgroundAccessStatus.DeniedBySystemPolicy)
    End Function

    Public Function RequestAccessAsync() As IAsyncOperation(Of Boolean)
        Return RequestAccessImplAsync().AsAsyncOperation()
    End Function

    Private Sub UnregisterTask(name As String)
        For Each v In From t In BackgroundTaskRegistration.AllTasks
                      Let r = t.Value Where r.Name = name Select r
            v.Unregister(True)
        Next
    End Sub

    Private Sub RegisterTask(name As String, type As Type, Optional trigger As IBackgroundTrigger = Nothing, Optional condition As IBackgroundCondition = Nothing)
        Dim task As New BackgroundTaskBuilder
        task.Name = name
        task.TaskEntryPoint = type.FullName
        If trigger IsNot Nothing Then task.SetTrigger(trigger)
        If condition IsNot Nothing Then task.AddCondition(condition)
        task.Register()
    End Sub

    Private Const LIVETILETASK As String = "LIVETILETASK"

    Public Sub RegisterLiveTile(reg As Boolean)
        UnregisterTask(LIVETILETASK)
        If reg Then
            RegisterTask(LIVETILETASK, GetType(LiveTileTask), New TimeTrigger(15, False), New SystemCondition(SystemConditionType.InternetAvailable))
        End If
    End Sub

    Private Const LOGINTASK As String = "LOGINTASK"

    Public Sub RegisterLogin(reg As Boolean)
        UnregisterTask(LOGINTASK)
        If reg Then
            RegisterTask(LOGINTASK, GetType(LoginTask), New SystemTrigger(SystemTriggerType.NetworkStateChange, False))
        End If
    End Sub
End Module
